### ISSUE TITLE
DateTimeConverter transforms throws an InvalidCastException when a null value is passed in

### ISSUE THREAD

**Issue by mwassell** [lithnet/umare#1]
Attempting to create an attribute flow rule to do the following:
1. Lookup Value - If value is found on import, replace with NULL. If value is not found, return value.
2. Convert to DateTime - Convert ".NET ticks" to DateTime fails and returns an error message in sync engine if value is NULL.

Example:
accountExpires>>LookupValue>>ConvertToDateTime>>accountExpires

Thank you!

**Comment by ryannewington** [comment]
Fixed transform library and updated nuget package https://github.com/lithnet/transform-engine/commit/288673c34e4bb4e91d0e71cd4e1e211dffed46f1

**Comment by ryannewington** [comment]
@mwassell, 

Test this release and see if it resolves the issue for you 
https://github.com/lithnet/umare/releases/tag/v1.5.6117

Let me know how it goes.

Ryan

**Comment by mwassell** [comment]
@ryannewington Thank you for the quick response Ryan! 

This flow doesn't seem to work:
accountExpires>>LookupValue>>ConvertToDateTime>>accountExpires

It appears that LookupValue returns a NULL of type String by default. 

But, this flow does work:
accountExpires>>LookupValue>>ToInteger>>ConvertToDateTime>>accountExpires

**Comment by ryannewington** [comment]
Can you attach or email me your umarex file? Also, the exact error message you are getting?

**Comment by mwassell** [comment]
Sorry for the delay! I've attached the UMAREX transform file. Here's the error:

```
System.FormatException: A null object cannot be converted to 'long'
   at Lithnet.MetadirectoryServices.TypeConverter.ConvertToLong(Object obj)
   at Lithnet.MetadirectoryServices.TypeConverter.ConvertData(Object obj, ExtendedAttributeType type)
   at Lithnet.Transforms.SimpleLookupTransform.TransformSingleValue(Object inputValue)
   at Lithnet.Transforms.Transform.TransformValue(Object inputValue)
   at Lithnet.Transforms.Transform.TransformValue(IList`1 inputValues)
   at Lithnet.Transforms.Transform.ExecuteTransformChain(IEnumerable`1 transforms, IList`1 inputValues)
   at Lithnet.Umare.MAExtensionObject.Microsoft.MetadirectoryServices.IMASynchronization.MapAttributesForImport(String flowRuleName, CSEntry csentry, MVEntry mventry)
```

[umarex-transforms.txt](https://github.com/lithnet/umare/files/507032/umarex-transforms.txt)

**Comment by ryannewington** [comment]
@mwassell If I understand your logic correctly, you want to take account expires values of 'never' in AD and have them map into something useful in FIM. I can see that the value for account expires 'never' is Int64.MaxValue in AD, but that exceeds the allowed value for FILETIME (the underlying data type in AD) which the transform uses for AD dates. If this is the case, i'm wondering if we actually make a special case for handling this in the date converter transform. (The 'filetime' input type is actually designed to handle native AD date data - it seems we don't handle this situation).

The date would come through as 9999-12-31, which you can then do a lookup transform on and replace it with null if you didnt want that value?

Would that work for you? Then you could just do something like

accountExpires>>ConvertFromFileTime>>accountExpires

or if you didnt want the 9999-12-31 value

accountExpires>>ConvertFromFileTime>>ReplaceMaxDateWithNull>>accountExpires

**Comment by mwassell** [comment]
@ryannewington Correct! Thank you for taking the time to understand this situation.

The eventual goal is to export the value to the Portal. I thought that replacing the invalid value with a Null and not allowing Nulls to be Exported will provide the desired result.

**Comment by ryannewington** [comment]
Ok try this version.
If you create a new DateTimeConverter transform with 'FileTime' type selected as the input type, and FIM Service Date as the output type, it will convert those values to 9999-12-31.

accountExpires>>ConvertFromFileTime>>accountExpires

[Lithnet.Umare.msi.zip](https://github.com/lithnet/umare/files/512507/Lithnet.Umare.msi.zip)

**Comment by mwassell** [comment]
Thanks Ryan! I'll give it a go and let you know the results,

**Comment by ryannewington** [comment]
Hi @mwassell, 

Did you have a chance to look at this?

Ryan

**Comment by mwassell** [comment]
You are the man! It works perfectly.

Thank you for all of your hard work (again). :-)

**Comment by ryannewington** [comment]
Released as production version 1.5.6123

https://github.com/lithnet/umare/releases/tag/v1.5.6123
